### ISSUE TITLE
[Feature Request] llama.cpp connection

### ISSUE THREAD

**Issue by LiliumSancta** [Cohee1207/SillyTavern#371]
Would it be possible to add direct connection to llama.cpp? This week a pr was added with an example of a server that has basic endpoints for generation without the need for bindings like python or even the ooba api. The biggest advantage in all of this is that the generation speed of llama.cpp has always been faster than the bindings created on top of the llama.cpp c api. 

I know there are issues with context, regenerating responses, etc. But I thought it would be interesting for the speed.

https://github.com/ggerganov/llama.cpp/tree/master/examples/server

**Comment by gustrd** [comment]
The easiest way is to use KoboldCpp instead of LlamaCpp. It has a minimal KobolAI API that allows connecting with SillyTavern, and it`s a fork of LlamaCpp being constantly updated.

https://github.com/LostRuins/koboldcpp

**Comment by LiliumSancta** [comment]
> The easiest way is to use KoboldCpp instead of LlamaCpp. It has a minimal KobolAI API that allows connecting with SillyTavern, and it`s a fork of LlamaCpp being constantly updated.
> 
> https://github.com/LostRuins/koboldcpp

Yes, I know koboldcpp, but even koboldcpp won't provide the speed of llama.cpp especially in streaming and perhaps not even what I suggested can.

**Comment by LiliumSancta** [comment]
> It is possible to even include llama-node into the mix for an in-house generation.: https://github.com/Atome-FE/llama-node However, adding new API connections currently has a low priority. Will consider that for potential future updates.

I confess that I haven't tested llama-node or go-llama, with any frontend. 
I believe this llama.cpp server could become more complete in the future and maybe frontends and clients will use it directly when possible. Thanks for the answer and I think I'll try it via proxy for now using Sillytavern.

**Comment by LiliumSancta** [comment]
> It is possible to even include llama-node into the mix for an in-house generation.: https://github.com/Atome-FE/llama-node However, adding new API connections currently has a low priority. Will consider that for potential future updates.

https://github.com/mlc-ai/web-llm/commit/8ce1d8d8304c90c779a4cc72433993a0d6c797e4
Just leaving this here since you mentioned the possibility of having an in-house generation solution in the future, at the moment it only uses of gpu through webgpu which limits models with possible size for vram, but it is cross-platform and has respectable speed.
As it is not a priority at the moment I think this can be closed.

**Comment by digiwombat** [comment]
The llama.cpp api server example is a thing now. I have a PR out to make it work as expected with frontends (it currently doesn't account for anything that edits context), though I guess we'll see if it gets merged. It's about twice as fast as the python alternative. I assume it would be faster than the node/llama.cpp interop as well.

If anyone wants to test it out, it is currently running with simple-proxy if you grab my PR on llama.cpp (or just copy paste the code into the latest pull) and build the server from that. It's currently my go-to since running 30B on a 12GB card is very usable now, especially with the recent speedups.

**Comment by ericskiff** [comment]
Following up on this issue, I believe this is now working and complete. I am running llama.cpp with the --api flag, I load the model via the llama.cpp web UI, and then connect SillyTavern to http://127.0.0.1:5000/api and ws://127.0.0.1:5005/api/v1/stream. 

It works amazingly well, especially on my mac where llama.cpp has strong Metal support and handles longer contexts and llama2 perfectly

**Comment by FitzWM** [comment]
> Following up on this issue, I believe this is now working and complete. I am running llama.cpp with the --api flag, I load the model via the llama.cpp web UI, and then connect SillyTavern to http://127.0.0.1:5000/api and ws://127.0.0.1:5005/api/v1/stream.

What do you mean by this? I don't see any --api switch for llama.cpp on my system or in the documentation, and trying to run it with that switch gives me an error. I know Ooba has a switch like that, but I'm not finding one with llama.cpp. I'd love to be wrong, though!

**Comment by ericskiff** [comment]
Oh my gosh. You're absolutely right. I got this working a few nights ago and forgot I got it working by going through oobabooga. Apologies, disregard my comment!

**Comment by RDearnaley** [comment]
It would be really nice to get this working, and be able to use the main llama.cpp project with its latest updates, rather than koboldcpp or one of a variety of other hacky/slow workarounds.

I'm not sure that I agree with the LowPriority assessment - it seems to me that in a fast-changing field like AI, a front-end like Silly Tavern should interoperate with as many major back ends as possible, particularly including ones that are easy to set up, well maintained, support the latest open-source models, and a good way to run large models on consumer hardware, as llama.cpp is.

**Comment by Cohee1207** [comment]
I'm not sure I'd be up to speed with maintaining both the frontend and backend in the same project.
And I generally don't like llama.cpp, but if you want something that isn't kobold, but uses the same GGML backend, check out these projects that should be compatible with ST:

1. [LM Studio](https://lmstudio.ai/)
2. [Local AI](https://www.localai.app/)
3. I'm sure there's more that I'm not aware of

**Comment by digiwombat** [comment]
I'm sure LostRuins could probably speak to this more thoroughly, but the speed difference with native and koboldcpp (and llama-cpp-python which ooba's webui uses) is much less drastic than it was previously.

If you want to run the server example with SillyTavern no matter what, you can also take a look at simple-proxy-for-tavern which works with the llama.cpp server.exe and writing an essentially blank passthrough preset for that should be pretty simple. Tavern's instruct formatting makes that mostly obsolete, but it's still a useful tool if you want to do work with server.exe.

**Comment by Cohee1207** [comment]
Also, you'd need to use bigger token paddings with koboldcpp since there's something wrong with token counting even when using Llama's native tokenizer. Values around 200 should be okay for the 4096 context.

**Comment by geoffsmithBK** [comment]
Looks like llama.cpp's server is now supporting a fully native OAI api, exporting endpoints like /models, /v1/{completions, chat/completions}, etc. Essentially implementing the old 'simple-proxy-for-tavern' functionality and so can be used with ST directly w/o api_like_OAI.py, kcpp, ooba, etc. Works great this way and is nice and fast, similar pp/gen times to llama.cpp native main and server:

https://github.com/ggerganov/llama.cpp/pull/4198

https://github.com/ggerganov/llama.cpp/issues/4216

https://www.reddit.com/r/LocalLLaMA/comments/185kbtg/llamacpp_server_rocks_now/

Maybe now's a good time to look at this again? Or how could someone add some of the newer open source hyperparams to the api calls (e.g. mirostat, min_p, tfs_z, etc.)? Doesn't look like this can be done just in the .settings file or can it?

**Comment by Cohee1207** [comment]
These params won't be sent unless added as a new Chat Completion or Text Completion source.

**Comment by geoffsmithBK** [comment]
> These params won't be sent unless added as a new Chat Completion or Text Completion source.

Got it. Seems straightforward, I gather the disinclination is the additional point of maintenance?

**Comment by Cohee1207** [comment]
Added to staging.
